### PR TITLE
Add deletion buttons to detail view for application, teachercon reg, …

### PR DESCRIPTION
**Pull request by clareconstantine**
…and fit wknd reg

delete menu options:
screen shot 2018-04-24 at 5 46 07 pm

confirmation dialog:
screen shot 2018-04-19 at 3 53 49 pm

### CODE REVIEW

**Comment by clareconstantine** [comment]
[Let Code.org admins delete applications/registrations](https://trello.com/c/CvHloKBw/567-let-codeorg-admins-delete-applications-registrations)

File: apps/src/code-studio/pd/application_dashboard/detail_view_contents.jsx [hunk 1]
```diff
@@ -361,6 +423,63 @@ export class DetailViewContents extends React.Component {
...
+            </Button>
+            <ConfirmationDialog
+              show={this.state.showDeleteApplicationConfirmation}
+              onOk={this.handleDeleteApplicationConfirmed}
+              onCancel={this.handleDeleteApplicationCancel}
+              headerText="Delete Application"
+              bodyText="Are you sure you want to delete this application? You will not be able to undo this."
+              okText="Delete"
+            />
+          </div>
```
Comment (aoby) [inline apps/src/code-studio/pd/application_dashboard/detail_view_contents.jsx]: Why is this all being added to `renderLockButton`? That name no longer describes the function. I'd prefer to see these in their own functions, or maybe a combined function for all of them, but in any case they don't seem related to the lock button.

File: apps/src/code-studio/pd/application_dashboard/detail_view_contents.jsx [hunk 2]
```diff
@@ -227,6 +236,59 @@ export class DetailViewContents extends React.Component {
...
+    this.setState({showDeleteTeacherconRegistrationConfirmation: false});
+  };
+
+  handleDeleteTeacherconRegistrationConfirmed = () => {
+    $.ajax({
+      method: "DELETE",
+      url: `/pd/teachercon_registration/${this.props.applicationData.application_guid}`
+    }).done(() => {
+      this.setState({showDeleteTeacherconRegistrationConfirmation: false});
+    });
```
Comment (aoby) [inline apps/src/code-studio/pd/application_dashboard/detail_view_contents.jsx]: note: the error case is handled above, but not here and not below. Do these registration deletes have any feedback other than the delete button disappearing?

**Comment by mehalshah** [comment]
Agree with Andrew's comment - hide them under the admin menu

**Comment by mehalshah** [comment]
Also agree that we should only have this button to workshop admins and not group 3 partners.

File: apps/src/code-studio/pd/application_dashboard/detail_view_contents.jsx (hunk 1 above)
Comment (clareconstantine) [reply apps/src/code-studio/pd/application_dashboard/detail_view_contents.jsx]: oh, they aren't - they are in`renderEditMenu`. The collapsed view on github is misleading.

**Comment by mehalshah** [comment]
I think that if we are going to give workshop admins delete permission, it's time to make applications paranoid. @aoby / @asherkach if there's nothing more to it then adding `acts_as_paranoid` to `Pd::Application::ApplicationBase` then lets do it with this change.

File: apps/test/unit/code-studio/pd/application_dashboard/detail_view_contentsTest.js [hunk 3]
```diff
@@ -150,6 +152,42 @@ describe("DetailViewContents", () => {
...
+      it("Does not have delete registration menu items if there are not registrations", () =>{
+        const detailView = mountDetailView(applicationData.type, {isWorkshopAdmin: true});
+        detailView.find("#admin-edit").first().simulate("click");
+        const deleteTeacherconRegistrationMenuitem = detailView.find(".dropdown.open a").findWhere(a => a.text() === "Delete Teachercon Registration");
+        const deleteFitWeekendRegistrationMenuitem = detailView.find(".dropdown.open a").findWhere(a => a.text() === "Delete FiT Weekend Registration");
+
+        expect(deleteTeacherconRegistrationMenuitem).to.have.length(0);
+        expect(deleteFitWeekendRegistrationMenuitem).to.have.length(0);
+      });
     });
```
Comment (mehalshah) [inline apps/test/unit/code-studio/pd/application_dashboard/detail_view_contentsTest.js:191]: I would like to see a test verifying that these buttons do not appear for non-workshop-admins

**Comment by aoby** [comment]
@mehalshah I think that's a great idea. We also need to a `deleted_at` column.

File: apps/test/unit/code-studio/pd/application_dashboard/detail_view_contentsTest.js (hunk 3 above)
Comment (clareconstantine) [reply apps/test/unit/code-studio/pd/application_dashboard/detail_view_contentsTest.js:191]: We have a test that this whole menu doesn't appear for non-workshop admins: https://github.com/code-dot-org/code-dot-org/blob/ce01f3fe4b22c586b116784877b89ce8593e1662/apps/test/unit/code-studio/pd/application_dashboard/detail_view_contentsTest.js#L132
